File: queue/0239_hard/window_maximums.py

def find_window_maximums(nums: list[int], window_size: int) -> list[int]:  # LeetCode Q.239.
    window_maximums = []
    # Queue contains indices of nums under current sliding window.
    queue = []  # Leftmost idx always points to window maximum.
    for idx, num in enumerate(nums):
        # Ensure last idx's num > current num.
        while queue and nums[queue[-1]] <= num:
            queue.pop(-1)

        # Ensure leftmost idx is in window.
        while queue and idx - window_size + 1 > queue[0]:
            queue.pop(0)

        queue.append(idx)
        if idx >= window_size - 1:  # Can collect rolling window maximums.
            window_maximums.append(nums[queue[0]])

    return window_maximums

File: queue/0239_hard/window_maximums.cpp
#include <deque>
#include <vector>
using namespace std;

vector<int> find_window_maximums(vector<int> &nums, int window_size)
{ // LeetCode Q.239.
    vector<int> window_maximums;
    // Nums decreasing monotonic queue. Format: (num, idx).
    deque<pair<int, int>> queue;

    for (int idx = 0; idx < nums.size(); idx++)
    {
        while (!queue.empty() && queue.front().second + window_size <= idx)
        {
            queue.pop_front(); // Max num is no longer in current window.
        }
        while (!queue.empty() && queue.back().first < nums[idx])
        {
            queue.pop_back(); // Bigger nums replace smaller nums.
        }
        queue.push_back({nums[idx], idx});

        if (idx >= window_size - 1)
        { // Can collect rolling window maximums.
            window_maximums.push_back(queue.front().first);
        }
    }

    return window_maximums;
}
